### PR TITLE
Examples of proposed approach toward eliminating dts_fixup for device drivers

### PR DESCRIPTION
**Pull request by pabigot**
Discussion in and around #11737 has reached the point where the remaining approach that does not require significant tooling is to use an alias for each driver instance.  This PR shows what that might look like for two cases:
* the sole example in the baseline of a non-SOC device with multiple instances (change to proposed standard prefix instead of old aliases)
* a sensor device with trigger support

For the latter the patch also adds the config structure required to support for multiple instances, though the driver currently instantiates only one.

The patch set also contains an alternative to, and documents proposed removal of, the generated identifier prefix `DT_compatible_alias_`, instead prefering `DT_alias` for reasons raised in multiple discussions linked to from the first commit message.

And it includes a fix for test failures in an untouched eswifi module; I don't know how that got past CI in the first place.

Please provide feedback on whether this approach is acceptable as standard practice.

### CODE REVIEW

File: samples/sensor/sht3xd/nrf51_ble400.overlay [hunk 1]
```diff
@@ -16,3 +16,9 @@
 		alert-gpios = <&gpio0 2 GPIO_INT_ACTIVE_HIGH>;
 	};
 };
+
+/ {
+	aliases {
+		sht3xd-0 = &sht3xd;
```
Comment (erwango) [inline samples/sensor/sht3xd/nrf51_ble400.overlay]: IIUC, aim of this alias is to instantiate a generic set of defines that could be used to implement  samples/sensor/sht3xd, right?

**Comment by erwango** [comment]
Seems consistent approach. This way of using alias allows for generic code implementation.

File: samples/sensor/sht3xd/nrf51_ble400.overlay (hunk 1 above)
Comment (pabigot) [reply samples/sensor/sht3xd/nrf51_ble400.overlay]: Yes.

**Comment by mnkp** [comment]
The comment to the commit "scripts: extract: remove compatible component from generated alias" states
```
The rationale for encoding compatible identifier in the alias prefix was
in support of a feature (Kconfig-based identification of compatible
drivers) that has not yet been designed/implemented.
```
Could you explain a bit more how this Kconfig-based identification was supposed to be implemented?

At the moment we generate 3 different versions of alias defines. We need to agree on one and remove the others.

**Comment by pabigot** [comment]
I can explain no more about the Kconfig-based identification plan than whatever you take in the comments referenced in the commit messages, which was where I found it referenced.

I agree on removing two of the three aliases.  Doing so is not in scope of this example and should probably be discussed in 11737.  What I hope for here is reasons why this approach should **not** be taken, based on evidence of inferior maintainability and expressibility when compared with fully-described alternative solutions.

If none are found, I will propose we run with it and deprecate the others.

**Comment by mnkp** [comment]
As for the "scripts: extract: remove compatible component from generated alias" commit. Unless there are voices against, in the past @galak preferred to have compatible component added, maybe we could prepare a separate PR which generates only one type of define for an alias, the one with DT_ prefix and compatible removed, and cleans up all other use cases. So far only Nordic SoCs rely on compatible component in the alias. And the old unprefixed aliases are due to be removed anyway. I will be happy to help.

**Comment by carlescufi** [comment]
@anangl @Mierunski could you please take a look at the proposal and give your opinion based on the work we've already done on aliases with Nordic?

**Comment by galak** [comment]
> As for the "scripts: extract: remove compatible component from generated alias" commit. Unless there are voices against, in the past @galak preferred to have compatible component added, maybe we could prepare a separate PR to generate only one type of define for an alias, the one with DT_ prefix and compatible removed, and clean up all other use cases. So far only Nordic SoCs rely on compatible component in the alias. And the old unprefixed aliases are due to be removed anyway. I will be happy to help.

I think it would be good to see what the implication of removing the unprefix'd aliases would be.  I'm guessing this might impact a few samples, but its worth cleaning them up.  I'd do that as a first step.

**Comment by pabigot** [comment]
@galak I believe the unmerged NRF91 material may use the aliases that include a compatible component, since that's all that was available to them.  So we'll probably have to delay until that's merged.

How do we drive this to conclusion?  Do you want the removal of at least the `DT_compatible_foo` aliases to be part of this PR, or do you still feel there's some use case justifying retaining them?  Do you see any lingering questions about whether this approach is the way to go?

**Comment by derchrismakaio** [comment]
@pabigot  as discussed in #12088 I feel the same pain as you do. 
Please confirm if I got your proposed procedure right for my new driver.

The driver is for max170xx fuel gauge family, currently I have planned for max17043, max17045 and max17055. I don't think that there will be multiple instances running, so I decided to "alias" like this. max170xx = &max170xx;

here's how my custom board dts looks like:

```
&i2c1 {
	status = "ok";
    sda-pin = <24>;
    scl-pin = <41>;

    max170xx:max17043 {
        label = "max17043";
        compatible = "max170xx";
        status = "ok";
    };
};

/ {
	aliases {
		max170xx = &max170xx;
	};
};
```

and this gives me the nice definitions I really like:
```
#define MAX170XX_BUS_NAME    DT_NORDIC_NRF_I2C_40004000_MAX170XX_40004000_MAX17043_BUS_NAME
#define MAX170XX_LABEL   DT_NORDIC_NRF_I2C_40004000_MAX170XX_40004000_MAX17043_LABEL
```

is this how you proposed it?

**Comment by galak** [comment]
> How do we drive this to conclusion? Do you want the removal of at least the `DT_compatible_foo` aliases to be part of this PR, or do you still feel there's some use case justifying retaining them? Do you see any lingering questions about whether this approach is the way to go?

So I'm coming back to this, and looking at it from a fresh perspective.  I feel like we should use `DT_compatible_foo` as what we support for sensor drivers like i2c/spi.  I don't think use of DTS aliases makes sense for those.  Since its only the EEPROM test that has issues, we can either rework the test to only use one EEPROM for now.

**Comment by pabigot** [comment]
@galak So as far as I understand it, that's a NACK on this approach.  We won't use aliases to provide standard names for subordinate devices, so we'll continue to require dts_fixup.h files to convert from the generated symbols that include SOC-specific bus paths to a generic symbol that can be used in drivers until the device tree tooling and code generation support is complete.

Do I have that right?

**Comment by chris-makaio** [comment]
@pabigot just my 2ct, but please don't remove the aliases. For now it's the only approach to get some board-independent app<->HAL code

**Comment by galak** [comment]
> @galak So as far as I understand it, that's a NACK on this approach. We won't use aliases to provide standard names for subordinate devices, so we'll continue to require dts_fixup.h files to convert from the generated symbols that include SOC-specific bus paths to a generic symbol that can be used in drivers until the device tree tooling and code generation support is complete.
> 
> Do I have that right?

No, I'm saying do what I did in PR #11886 for I2C/SPI devices.

**Comment by galak** [comment]
I'm saying we take a 2 prong approach.  For SoC devices we can use aliases like is being done for Nordic & Silabs.

For non-SoC (i2C & SPI) devices we'd use a simplified compatible based define, with the caveat that you can only have one such device in the system for the LTS release.

**Comment by mnkp** [comment]
Why we don't want to use aliases for non-SoC devices? What's the reasoning behind this choice?

**Comment by pabigot** [comment]
This PR is specifically intended to eliminate the need for dts_fixup.h for I2C/SPI devices, using the existing practice of aliases in the same way as soc peripherals so everything's consistent.

As I commented in #11886 relying on compatible alone doesn't allow multiple devices: the test case that fails is based on having two devices, and I think that's a valid configuration that should be allowed.

**Comment by chris-makaio** [comment]
I don‘t get why we don‘t simply use instance numbers in the aliases.
I mean, it‘s basically the same for the „on SoC“ devices like UART0 and UART1

**Comment by pabigot** [comment]
From @galak (personal communication):
> the issue I have with aliases is that we then have to define an alias name space across all the sensors, and we already have one in place with the compatible

It goes beyond sensors as there are power, GPIO, storage, and other devices for which device tree entries must map to identifiers in drivers and user code.  There aren't so many devices that uniqueness is difficult (the name of the driver as reflected in its containing directory or file name is an obvious choice), but I do have some sympathy with this position.

At this time conflicting proposals still exist: leave things in their current state; use aliases as proposed here; or use the solution in #11886.    Which needs review, so if anybody has interest in this please go comment there.

**Comment by pabigot** [comment]
Closing this in favor of a revised version that will be based on #12426 (or its successor if we're still churning).

**Pull request closed without merging**